### PR TITLE
ci(sort_cspell_json): use pre-commit

### PR DESCRIPTION
**Pull request by KeisukeShima**
Changed to use `pre-commit.ci` so that git push can be used in the forks.
TODO: After this PR is merged, create a tag `v0.1.0`.

Tested on #253 

Signed-off-by: Keisuke Shima <19993104+KeisukeShima@users.noreply.github.com>

### CODE REVIEW

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [inline pre_commit_hooks/sort_cspell_json.py:50]:
```suggestion
```

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [inline pre_commit_hooks/sort_cspell_json.py:28]: Adding type hints is better.

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [inline pre_commit_hooks/sort_cspell_json.py:63]: I feel this can be simplified.

```suggestion
        sorted_json = parse(original_json)

        if json.dumps(sorted_json) != json.dumps(original_json):
```

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [inline pre_commit_hooks/sort_cspell_json.py:57]: Is it possible to write like this and remove this conversion? :thinking: 

```py
parser.add_argument("filenames", nargs="*", type=List[Path], help="Filenames to fix")
```

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [inline pre_commit_hooks/sort_cspell_json.py:61]: This function should be named `format_cspell_json` or `sort_cspell_json`.

```suggestion
        sorted_json = format_cspell_json(original_json)
```

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [inline pre_commit_hooks/sort_cspell_json.py:40]: For more readability.

```suggestion
def format_cspell_json(cspell_json):
```

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [inline pre_commit_hooks/sort_cspell_json.py:45]: Could you briefly explain this logic?

File: package.json
Comment (KeisukeShima) [reply package.json:2]: > Is this necessary? Seeing [tier4/pre-commit-hooks-ros#20 (comment)](https://github.com/tier4/pre-commit-hooks-ros/pull/20#issuecomment-1049830255), it's for Prettier and I think we don't use Prettier.

I see. Then, I will remove this file.

Comment (KeisukeShima) [reply package.json]: fbcd66f

Comment (kenji-miyake) [reply package.json]: Did it work correctly?

File: .github/workflows/pytest.yaml
Comment (KeisukeShima) [reply .github/workflows/pytest.yaml:15]: Should I correct below link as well?
https://github.com/tier4/pre-commit-hooks-ros/blob/main/.github/workflows/pytest.yaml#L13

Comment (kenji-miyake) [reply .github/workflows/pytest.yaml:15]: It's up to you, but I feel it's better!

**Comment by KeisukeShima** [comment]
`.cspell.json` includes unknown words?? :confused: 

https://github.com/tier4/autoware-spell-check-dict/runs/5675471711?check_suite_focus=true
```
Warning: Unknown word (planing)
tests/test_sort_cspell_json/.cspell.answer.json:162:6 Unknown word (planing)
Warning: Unknown word (planing)
tests/test_sort_cspell_json/.cspell.input.json:162:6 Unknown word (planing)
```

File: package.json
Comment (KeisukeShima) [reply package.json]: Yes.

I modified `pre-commit-config.yaml`  for testing:
```
  - repo: https://github.com/KeisukeShima/autoware-spell-check-dict
    rev: fbcd66fa6b3cf564062c4d2db01e13bf5bfbfb2a
    hooks:
      - id: sort-cspell-json
```

And the result is:
```
❯ pre-commit try-repo . -a -v
[WARNING] Creating temporary repo with uncommitted changes...
[INFO] Initializing environment for /tmp/tmpdcczijjr/shadow-repo.
===============================================================================
Using config:
===============================================================================
repos:
-   repo: /tmp/tmpdcczijjr/shadow-repo
    rev: 3adf84249ed830bb6ea3541aeda57c24955ef327
    hooks:
    -   id: sort-cspell-json
===============================================================================
[INFO] Installing environment for /tmp/tmpdcczijjr/shadow-repo.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
sort cspell.json.........................................................Passed
- hook id: sort-cspell-json
- duration: 0.14s
- ```

**Comment by kenji-miyake** [comment]
@KeisukeShima It's an exception (to detect misspelling of `planning`). For `.cspell.json`, it's ignored.
https://github.com/tier4/autoware-spell-check-dict/blob/e4be3ec0fd6af74ac6df4328c2d2a9526d81608e/.cspell.json#L126

I guess you can remove the part from the input/answer.

File: pre_commit_hooks/sort_cspell_json.py
Comment (KeisukeShima) [reply pre_commit_hooks/sort_cspell_json.py]: 2ab9b60

File: pre_commit_hooks/sort_cspell_json.py
Comment (KeisukeShima) [reply pre_commit_hooks/sort_cspell_json.py]: 6157ca3

File: pre_commit_hooks/sort_cspell_json.py
Comment (KeisukeShima) [reply pre_commit_hooks/sort_cspell_json.py]: 6157ca3

File: pre_commit_hooks/sort_cspell_json.py
Comment (KeisukeShima) [reply pre_commit_hooks/sort_cspell_json.py]: I use the following dict.
There are three option for content: unique, sort, and key_order.
If unique is true, unique_json is performed on the corresponding key value.
If sort is true, sort_json is performed on the corresponding key value. key_order allows you to specify the order during sorting.
```
KEYS = {
    "ignorePaths": {"unique": True, "sort": True, "key_order": None},
    "words": {"unique": True, "sort": True, "key_order": str.lower},
}
```

File: pre_commit_hooks/sort_cspell_json.py
Comment (KeisukeShima) [reply pre_commit_hooks/sort_cspell_json.py]: 884ad11

File: pre_commit_hooks/sort_cspell_json.py
Comment (KeisukeShima) [reply pre_commit_hooks/sort_cspell_json.py]: 571e26d

File: tests/test_sort_cspell_json.py
Comment (KeisukeShima) [reply tests/test_sort_cspell_json.py]: How about this change?
fb652dd

**Comment by KeisukeShima** [comment]
> @KeisukeShima It's an exception (to detect misspelling of `planning`). For `.cspell.json`, it's ignored.
> 
> https://github.com/tier4/autoware-spell-check-dict/blob/e4be3ec0fd6af74ac6df4328c2d2a9526d81608e/.cspell.json#L126
> 
> I guess you can remove the part from the input/answer.

Okay, removed f4f22b0

File: tests/test_sort_cspell_json.py
Comment (kenji-miyake) [reply tests/test_sort_cspell_json.py]: What is `tmp_path`? :thinking: 
I think it's unnecessary.

Please see https://github.com/tier4/pre-commit-hooks-ros/blob/main/tests/test_ros_include_guard.py.

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [reply pre_commit_hooks/sort_cspell_json.py]: Also, could you change the file name to `format_cspell_json`, please?

File: tests/test_sort_cspell_json.py
Comment (kenji-miyake) [reply tests/test_sort_cspell_json.py]: https://github.com/tier4/pre-commit-hooks-ros/pull/54#discussion_r834897775

Comment (KeisukeShima) [reply tests/test_sort_cspell_json.py]: I understood. removed tmp_path.
3e360e1

Comment (kenji-miyake) [reply tests/test_sort_cspell_json.py]: @KeisukeShima Oh, then you'll fix https://github.com/tier4/pre-commit-hooks-ros/pull/54? :thinking:

File: pre_commit_hooks/sort_cspell_json.py
Comment (KeisukeShima) [reply pre_commit_hooks/sort_cspell_json.py]: a391926

File: tests/test_sort_cspell_json.py
Comment (KeisukeShima) [reply tests/test_sort_cspell_json.py]: @kenji-miyake In fact, there is no need to copy it.

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [reply pre_commit_hooks/sort_cspell_json.py]: Thank you!

File: tests/test_sort_cspell_json.py
Comment (kenji-miyake) [reply tests/test_sort_cspell_json.py]: Hmm? Could you explain a bit more in detail?

Comment (KeisukeShima) [reply tests/test_sort_cspell_json.py]: > pytest-datadir will copy the original file to a temporary folder, so changing the file contents won't change the original data file.
https://pypi.org/project/pytest-datadir/

Comment (kenji-miyake) [reply tests/test_sort_cspell_json.py]: Ah, I see.

Comment (KeisukeShima) [reply tests/test_sort_cspell_json.py]: Then, I changed test code.
d05fd58

Comment (KeisukeShima) [reply tests/test_sort_cspell_json.py]: > @KeisukeShima Oh, then you'll fix [tier4/pre-commit-hooks-ros#54](https://github.com/tier4/pre-commit-hooks-ros/pull/54)? thinking

Yes, I think it's better to simplify test code.

Comment (KeisukeShima) [reply tests/test_sort_cspell_json.py]: https://github.com/tier4/pre-commit-hooks-ros/pull/55

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [reply pre_commit_hooks/sort_cspell_json.py]: I've applied my suggestions. How about this? d6be47b1c88d5e5c11a71af416a1cd0641006223
I'm sorry for my drastic changes. :bow:

File: pre_commit_hooks/format_cspell_json.py
Comment (kenji-miyake) [inline pre_commit_hooks/format_cspell_json.py:41]: `languageSettings` is intentionally removed because it seemed unnecessary.
It just has `allowCompoundWords: false`, which is also defined in the top-level.

**Comment by kenji-miyake** [comment]
@KeisukeShima Could you check my changes, please? :pray:

File: pre_commit_hooks/sort_cspell_json.py
Comment (KeisukeShima) [reply pre_commit_hooks/sort_cspell_json.py]: * The results are different before and after your script change. Perhaps you changed the answer file for that.
* Why did you make a major change after the review was over? I do not understand why.

Comment (kenji-miyake) [reply pre_commit_hooks/sort_cspell_json.py]: > The results are different before and after your script change. Perhaps you changed the answer file for that.

Yes.

> Why did you make a major change after the review was over? I do not understand why.

What do you mean? :thinking: 
I think the review isn't over. It's my review and improvement.
If you want to split it into another PR, I can do it.

Comment (KeisukeShima) [reply pre_commit_hooks/sort_cspell_json.py]: Also, are the results consistent with the following commands in README?
https://github.com/tier4/autoware-spell-check-dict/blob/main/README.md?plain=1#L9

Comment (kenji-miyake) [reply pre_commit_hooks/sort_cspell_json.py]: > Also, are the results consistent with the following commands in README?

Yes, I think so.
But other keys are sorted as well.

File: pre_commit_hooks/format_cspell_json.py
Comment (KeisukeShima) [reply pre_commit_hooks/format_cspell_json.py:41]: Reference: https://github.com/streetsidesoftware/cspell/tree/main/packages/cspell#LanguageSettings

File: pre_commit_hooks/sort_cspell_json.py
Comment (KeisukeShima) [reply pre_commit_hooks/sort_cspell_json.py]: > > The results are different before and after your script change. Perhaps you changed the answer file for that.
> 
> Yes.
> 
> > Why did you make a major change after the review was over? I do not understand why.
> 
> What do you mean? thinking I think the review isn't over. It's my review and improvement. If you want to split it into another PR, I can do it.

I have made modifications according to your review. But putting in destructive changes after that review is problematic for the review process.
If you make any major changes in the future, please tell me before the review.
For the script, there is nothing wrong with the change, so I approve it.

**Comment by KeisukeShima** [comment]
@kenji-miyake Oh, I am the author of this PR, so I cannot approve it. Can you please merge them?

File: pre_commit_hooks/sort_cspell_json.py
Comment (kenji-miyake) [reply pre_commit_hooks/sort_cspell_json.py]: I'm sorry for my lack of explanation. :bow:

**Comment by kenji-miyake** [comment]
Yes, but pre-commit.ci is broken, I'll check it.

**Comment by kenji-miyake** [comment]
pre-commit.ci run

**Comment by KeisukeShima** [comment]
@kenji-miyake 
We need update tag after this pr merged.
TODO: After this PR is merged, create a tag v0.1.0.

**Comment by kenji-miyake** [comment]
Yeah, but I've replaced it with `repo: local`.

**Comment by kenji-miyake** [comment]
Oh, I understand, it's in your fork and I've pushed to the upstream.
Will re-push my change.

**Comment by kenji-miyake** [comment]
@KeisukeShima Could you confirm two new commits?

File: .cspell.json
Comment (kenji-miyake) [inline .cspell.json:162]: Oh, I should have added this.

Comment (kenji-miyake) [reply .cspell.json:162]: bd085cfb887e02f005ee7063e32c4b839f49c167

**Comment by KeisukeShima** [comment]
> @KeisukeShima Could you confirm two new commits?

Looks good to me :+1:

**Review by kenji-miyake — DISMISSED** [review]
@KeisukeShima Please merge if you think the changes are okay. :pray:

**Comment by kenji-miyake** [comment]
Ah, but pytest fails

**Comment by KeisukeShima** [comment]
> Ah, but pytest fails

Maybe the `flagWords` key is missing in `tests/test_format_cspell_json/.cspell.input.json`.

**Comment by kenji-miyake** [comment]
Yes, I'll fix it.

**Review by kenji-miyake — APPROVED** [review]

**Pull request merged**
